### PR TITLE
[Distributed] Support hierarchical partitioning

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
For multi-machine multi-GPU training, we can use hierarchical partitioning to increase the data locality between trainers. Each machine owns a first-level partition; each GPU trainer owns a second-level partition. The code is contributed by @soodoshll 

This optimization can reduce the number of input nodes in a mini-batch and reduce the data copy from CPU to GPU as well as improve the computation time.

Previous version on OGBN-products (2 machines, each machine has 4 GPUs):
```
part 0, train: 24577 (local: 23529), val: 4916 (local: 4620), test: 276637 (local: 268300)
part 1, train: 24577 (local: 23552), val: 4916 (local: 4632), test: 276637 (local: 268301)
part 2, train: 24577 (local: 23581), val: 4915 (local: 4666), test: 276636 (local: 268165)
part 3, train: 24577 (local: 23539), val: 4915 (local: 4608), test: 276636 (local: 268160)
part 4, train: 24577 (local: 24577), val: 4916 (local: 4916), test: 276637 (local: 276637)
part 5, train: 24577 (local: 24577), val: 4915 (local: 4915), test: 276636 (local: 276636)
part 6, train: 24577 (local: 24577), val: 4915 (local: 4915), test: 276636 (local: 276636)
part 7, train: 24576 (local: 24576), val: 4915 (local: 4915), test: 276636 (local: 276636)

Part 0, Epoch Time(s): 8.7029, sample+data_copy: 6.0583, forward: 1.2792, backward: 1.1113, update: 0.0213, #seeds: 24577, #inputs: 8637871
Part 1, Epoch Time(s): 8.7028, sample+data_copy: 5.1480, forward: 1.2844, backward: 2.0132, update: 0.0210, #seeds: 24577, #inputs: 8650060
Part 2, Epoch Time(s): 8.7031, sample+data_copy: 5.7049, forward: 1.2942, backward: 1.4449, update: 0.0211, #seeds: 24577, #inputs: 8651416
Part 3, Epoch Time(s): 8.7028, sample+data_copy: 6.5947, forward: 1.2694, backward: 0.5889, update: 0.0211, #seeds: 24577, #inputs: 8640628
Part 4, Epoch Time(s): 8.7029, sample+data_copy: 4.5067, forward: 1.2677, backward: 2.6642, update: 0.0215, #seeds: 24577, #inputs: 8296350
Part 5, Epoch Time(s): 8.7031, sample+data_copy: 4.8919, forward: 1.2437, backward: 2.3106, update: 0.0211, #seeds: 24577, #inputs: 8330937
Part 6, Epoch Time(s): 8.7028, sample+data_copy: 4.2798, forward: 1.3008, backward: 2.8142, update: 0.0211, #seeds: 24577, #inputs: 8282040
Part 7, Epoch Time(s): 8.7030, sample+data_copy: 4.0930, forward: 1.2712, backward: 3.0695, update: 0.0210, #seeds: 24576, #inputs: 8287232
```

With this optimization:
```
part 0, train: 24577 (local: 24577), val: 4916 (local: 4899), test: 276638 (local: 276481)
part 1, train: 24577 (local: 24577), val: 4916 (local: 4899), test: 276636 (local: 276480)
part 2, train: 24577 (local: 24577), val: 4915 (local: 4898), test: 276636 (local: 276480)
part 3, train: 24577 (local: 24577), val: 4915 (local: 4898), test: 276636 (local: 276480)
part 4, train: 24577 (local: 24531), val: 4916 (local: 4916), test: 276637 (local: 276637)
part 5, train: 24577 (local: 24531), val: 4915 (local: 4915), test: 276636 (local: 276636)
part 6, train: 24577 (local: 24531), val: 4915 (local: 4915), test: 276636 (local: 276636)
part 7, train: 24576 (local: 24531), val: 4915 (local: 4915), test: 276636 (local: 276636)

Part 0, Epoch Time(s): 6.8353, sample+data_copy: 4.6087, forward: 1.0405, backward: 0.9609, update: 0.0212, #seeds: 24577, #inputs: 6322639
Part 1, Epoch Time(s): 6.8353, sample+data_copy: 3.2540, forward: 0.9760, backward: 2.3869, update: 0.0223, #seeds: 24577, #inputs: 5711624
Part 2, Epoch Time(s): 6.8351, sample+data_copy: 4.2160, forward: 1.0694, backward: 1.3261, update: 0.0214, #seeds: 24577, #inputs: 6767261
Part 3, Epoch Time(s): 6.8351, sample+data_copy: 4.8576, forward: 1.0278, backward: 0.7436, update: 0.0209, #seeds: 24577, #inputs: 6288895
Part 4, Epoch Time(s): 6.8348, sample+data_copy: 4.9398, forward: 1.0306, backward: 0.6462, update: 0.0211, #seeds: 24577, #inputs: 6380756
Part 5, Epoch Time(s): 6.8351, sample+data_copy: 4.0851, forward: 1.0139, backward: 1.5213, update: 0.0214, #seeds: 24577, #inputs: 6068489
Part 6, Epoch Time(s): 6.8348, sample+data_copy: 4.5619, forward: 0.9839, backward: 1.0859, update: 0.0213, #seeds: 24577, #inputs: 5805581
Part 7, Epoch Time(s): 6.8350, sample+data_copy: 3.4411, forward: 0.9091, backward: 2.2817, update: 0.0218, #seeds: 24576, #inputs: 5131945
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/distributed/dist_graph.py [hunk 1]
```diff
@@ -1091,25 +1089,41 @@ def _split_even(partition_book, rank, elements):
...
+    return part_eles
 
-def node_split(nodes, partition_book=None, ntype='_N', rank=None, force_even=True):
+def _split_random_within_part(partition_book, rank, part_eles):
+    # If there are more than one client in a partition, we need to randomly select a subset of
+    # elements in the partition for a client. We have to make sure that the set of elements
+    # for different clients are disjoint.
+
+    num_clients = role.get_num_trainers()
+    num_client_per_part = num_clients // partition_book.num_partitions()
```
Comment (classicsong) [inline python/dgl/distributed/dist_graph.py:1107]: If num_clients is 6, partitions is 4, then num_client_per_part = 1.
Will this cause any problem?

**Review by classicsong — COMMENTED** [review]
There is no test code?

File: python/dgl/distributed/dist_graph.py (hunk 1 above)
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:1107]: this shouldn't happen. Each machine should have the same number of trainers. But we should add an assert to verify it.

Comment (classicsong) [reply python/dgl/distributed/dist_graph.py:1107]: Did you add the assertion in somewhere?

**Review by classicsong — APPROVED** [review]
LGTM.

**Pull request merged**